### PR TITLE
Detection status router merge queried segments #590

### LINKED ISSUE #590: Detection status router: merge queried segments
subtask for #504 

merge results of requested time ranges

### PR DESCRIPTION
**Pull request by amper43**
fixes #590

### CODE REVIEW

File: server/src/utils/spans.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+import * as _ from 'lodash';
+
+export function insertToSorted(array: number[], value: number) {
+  return array.splice(_.sortedIndex(array, value), 0, value);
+}
+
+export function getIntersectedDetections(detection: Detection, detections: Detection[], status?: DetectionStatus): Detection[] {
+  if(status === undefined) {
+    status = DetectionStatus.READY;
```
Comment (rozetko) [inline server/src/utils/spans.ts]: why is it `READY`?

File: server/src/services/data_puller.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import { AnalyticsTask, AnalyticsTaskType } from '../models/analytics_task_model';
 import * as AnalyticUnit from '../models/analytic_unit_model';
 import * as AnalyticUnitCache from '../models/analytic_unit_cache_model';
+import { mergeDetecionSpan } from '../controllers/analytics_controller';
```
Comment (rozetko) [inline server/src/services/data_puller.ts]:
```suggestion
import { mergeDetectionSpan } from '../controllers/analytics_controller';
```

Comment (rozetko) [reply server/src/services/data_puller.ts]: btw, why is it imported but not used?

File: server/src/controllers/analytics_controller.ts [hunk 3]
```diff
@@ -442,3 +446,69 @@ export async function runLearningWithDetection(id: AnalyticUnit.AnalyticUnitId)
...
+    });
+  }
+
+  let result: Detection[] = [];
+  intersectedDetections.map(i => result.push(new Detection(i.analyticUnitId, i.from, i.from, DetectionStatus.READY)));
+  newDetectionRanges.map(n => result.push(new Detection(n.analyticUnitId, n.from, n.from, DetectionStatus.RUNNING)));
+  return result;
+}
+
+export function mergeDetecionSpan(analyticUnitId: AnalyticUnit.AnalyticUnitId, from: number, to: number) {
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
export function mergeDetectionSpan(analyticUnitId: AnalyticUnit.AnalyticUnitId, from: number, to: number) {
```

File: server/src/controllers/analytics_controller.ts [hunk 4]
```diff
@@ -442,3 +446,69 @@ export async function runLearningWithDetection(id: AnalyticUnit.AnalyticUnitId)
...
+
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]: What should it do? Why is it empty?

File: server/src/controllers/analytics_controller.ts [hunk 5]
```diff
@@ -442,3 +446,77 @@ export async function runLearningWithDetection(id: AnalyticUnit.AnalyticUnitId)
...
+      .then(() => mergeDetecionSpan(analyticUnitId, from, to))
+      .catch(err => {
+        console.error(err);
+        _.find(detections, {analyticUnitId, from, to})[0].state = DetectionStatus.FAILED;
+      });
+    });
+  }
+
+  let result: Detection[] = [];
+  intersectedDetections.map(i => result.push(new Detection(i.analyticUnitId, i.from, i.from, DetectionStatus.READY)));
```
Comment (amper43) [inline server/src/controllers/analytics_controller.ts]: typo `from` & `from`

File: server/src/controllers/analytics_controller.ts [hunk 6]
```diff
@@ -442,3 +446,77 @@ export async function runLearningWithDetection(id: AnalyticUnit.AnalyticUnitId)
...
+    rangesBorders.push(d.to);
+  });
+  insertToSorted(rangesBorders, from);
+  insertToSorted(rangesBorders, to);
+
+  let alreadyDetected = false;
+  let startDetectionRange = null;
+  let newDetectionRanges: any[] = [];
+
+  for(let border of rangesBorders) {
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]: `utils`

File: server/src/utils/spans.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import { Detection, DetectionStatus } from '../models/detection_model';
+
+import * as _ from 'lodash';
+
+export function insertToSorted(array: number[], value: number) {
```
Comment (rozetko) [inline server/src/utils/spans.ts]: tests

File: server/src/controllers/analytics_controller.ts [hunk 8]
```diff
@@ -31,6 +33,8 @@ let dataPuller: DataPuller;
 
 let detectionsCount: number = 0;
 
+let detections: Detection[] = [];
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]: `TODO: save to db`

File: server/src/config.ts [hunk 9]
```diff
@@ -17,6 +17,7 @@ export const ANALYTIC_UNITS_DATABASE_PATH = path.join(DATA_PATH, 'analytic_units
 export const ANALYTIC_UNIT_CACHES_DATABASE_PATH = path.join(DATA_PATH, 'analytic_unit_caches.db');
 export const SEGMENTS_DATABASE_PATH = path.join(DATA_PATH, 'segments.db');
 export const THRESHOLD_DATABASE_PATH = path.join(DATA_PATH, 'treshold.db');
+export const DETECTION_STATUS_DATABASE_PATH = path.join(DATA_PATH, 'detection_status.db');
```
Comment (jonyrock) [inline server/src/config.ts]: let`s rename it to: 
```suggestion
export const DETECTIONS_DATABASE_PATH = path.join(DATA_PATH, 'detections.db');
```

File: server/src/models/detection_model.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+export enum DetectionStatus {
+  READY = 'READY',
+  RUNNING = 'RUNNING',
+  FAILED = 'FAILED'
+}
+
+export type DetectionId = string;
+
+export class Detection {
```
Comment (jonyrock) [inline server/src/models/detection_model.ts]: I suggest to rename type `Detection` to `DetectionSpan`

File: server/src/utils/spans.ts (hunk 1 above)
Comment (jonyrock) [reply server/src/utils/spans.ts]: also you can put default param here:

https://www.typescriptlang.org/docs/handbook/functions.html
see "Optional and Default Parameters" there

File: server/src/utils/spans.ts [hunk 11]
```diff
@@ -1,19 +1,60 @@
 import { Detection, DetectionStatus } from '../models/detection_model';
 
 import * as _ from 'lodash';
+import { toUnicode } from 'punycode';
+import { AnalyticUnitId } from '../models/analytic_unit_model';
```
Comment (jonyrock) [inline server/src/utils/spans.ts]: `AnalyticUnitId` in utils ?

File: server/src/controllers/analytics_controller.ts [hunk 12]
```diff
@@ -442,3 +447,45 @@ export async function runLearningWithDetection(id: AnalyticUnit.AnalyticUnitId)
     .then(() => runDetect(id))
     .catch(err => console.error(err));
 }
+
+export async function getDetectionSpans(analyticUnitId: AnalyticUnit.AnalyticUnitId, from: number, to: number): Promise<Detection[]> {
+
+  const unitCache = await AnalyticUnitCache.findById(analyticUnitId);
```
Comment (jonyrock) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
  const analyticUnitCache = await AnalyticUnitCache.findById(analyticUnitId);
```

File: server/src/models/analytic_unit_cache_model.ts [hunk 13]
```diff
@@ -28,6 +28,15 @@ export class AnalyticUnitCache {
       obj.data,
     );
   }
+
+  public getIntersection(): number {
```
Comment (jonyrock) [inline server/src/models/analytic_unit_cache_model.ts]: Very unclear by name `getIntersection` what method does

Comment (jonyrock) [reply server/src/models/analytic_unit_cache_model.ts]: also, this return type is not `number`, it more looks like `number | null`

File: server/spec/utils/spans.jest.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
+import { getNonIntersectedSpans } from '../../src/utils/spans';
+
+import 'jest';
+
+describe('getNonIntersectedSpans', function(){
```
Comment (rozetko) [inline server/spec/utils/spans.jest.ts]:
```suggestion
describe('getNonIntersectedSpans', function() {
```

**Pull request merged**